processors/baggage: add baggage span processor

codeboten opened the pull request:
Fixes #5397

pellared commented on processors/baggage/processor_test.go:
```diff
@@ -0,0 +1,63 @@
+	// create ctx with some baggage
+	ctx := context.Background()
+	suitcase := otelbaggage.FromContext(ctx)
+	packingCube, _ := otelbaggage.NewMember("baggage.test", url.PathEscape("baggage value"))
```
Use https://pkg.go.dev/go.opentelemetry.io/otel/baggage#NewMemberRaw instead.

codeboten replied:
Done 👍🏻

pellared commented on processors/baggage/processor_test.go:
```diff
@@ -0,0 +1,63 @@
+	// create ctx with some baggage
+	ctx := context.Background()
+	suitcase := otelbaggage.FromContext(ctx)
+	packingCube, _ := otelbaggage.NewMember("baggage.test", url.PathEscape("baggage value"))
```
Can you `require.NoError`?

codeboten replied:
Done 👍🏻

pellared commented on processors/baggage/processor_test.go:
```diff
@@ -0,0 +1,63 @@
+	ctx := context.Background()
+	suitcase := otelbaggage.FromContext(ctx)
+	packingCube, _ := otelbaggage.NewMember("baggage.test", url.PathEscape("baggage value"))
+	suitcase, _ = suitcase.SetMember(packingCube)
```
Can you `require.NoError`?

codeboten replied:
Done 👍🏻

pellared commented on processors/baggage/processor_test.go:
```diff
@@ -0,0 +1,63 @@
+
+func TestBaggageSpanProcessorAppendsBaggageAttributes(t *testing.T) {
+	// create ctx with some baggage
+	ctx := context.Background()
+	suitcase := otelbaggage.FromContext(ctx)
```
Can we use https://pkg.go.dev/go.opentelemetry.io/otel/baggage#New instead?

pellared commented on processors/baggage/processor_test.go:
```diff
@@ -0,0 +1,63 @@
+	_, span := tracer.Start(ctx, "test")
+	span.End()
+
+	assert.Equal(t, 1, len(exporter.spans))
+	assert.Equal(t, 1, len(exporter.spans[0].Attributes()))
```
You can use `assert.Len` instead.

codeboten replied:
Done

pellared commented on processors/baggage/processor_test.go:
```diff
@@ -0,0 +1,63 @@
+	assert.Equal(t, 1, len(exporter.spans))
+	assert.Equal(t, 1, len(exporter.spans[0].Attributes()))
+
+	for _, attr := range exporter.spans[0].Attributes() {
+		assert.Equal(t, attribute.Key("baggage.test"), attr.Key)
+		assert.Equal(t, "baggage value", attr.Value.AsString())
+	}
```
AFAIK you can simply

```suggestion
	want := []attribute.KeyValue{ attribute.String("baggage.test", "baggage value") }
	assert.Equal(t, want, exporter.spans[0].Attributes())
```

pellared commented on processors/baggage/processor.go:
```diff
@@ -0,0 +1,34 @@
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+type baggageSpanProcessor struct{}
```
Can we export the type so that we can possibly add additional methods later if needed?

codeboten replied:
@pellared we could, though I'm not sure if there's a need just yet. I don't have a strong preference one way or the other, happy to follow your guidance on this.

pellared replied:
> though I'm not sure if there's a need just yet

It is in order to avoid making a breaking change.

Lets rename it to


```suggestion
type SpanProcessor struct{}
```

codeboten replied:
👍🏻

codeboten replied:
Renamed and made public.

pellared commented:
Can you also update the `README.md`? I see this is the first processor. Shouldn't we have a "split" for `trace` and `log` processors?

codeboten commented:
@pellared i've updated the readme and moved the contents of the `baggage` package for traces under `baggage/baggagetrace` to follow the exporters pattern in the go repo.

dmathieu commented on CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add the new `go.opentelemetry.io/contrib/instrgen` package to provide auto-generated source code instrumentation. (#3068, #3108)
 - `NewSDK` in `go.opentelemetry.io/contrib/config` now returns a configured SDK with a valid `MeterProvider`. (#4804)
+- The `go.opentelemetry.io/contrib/processors/baggage` module.
+  This module provides a Baggage Span Processor. (#5397)
```
We're usually linking the PR doing the change, not the related issue.

```suggestion
  This module provides a Baggage Span Processor. (#5404)
```

dmathieu reviewed (commented):
How about adding an example_test?

pellared commented on processors/baggage/baggagetrace/doc.go:
```diff
@@ -0,0 +1,7 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package baggagetrace implements the baggage span processor which duplicates
+// onto a span the attributes found in Baggage in the parent context at
+// the moment the span is started.
```
This comment would better fit as a comment for the `SpanProcessor` type.

codeboten replied:
```suggestion
// Package baggagetrace implements a baggage span processor.
```

pellared reviewed (commented):
Just nit comments to address

Overall, LGTM

codeboten commented on CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Add the new `go.opentelemetry.io/contrib/instrgen` package to provide auto-generated source code instrumentation. (#3068, #3108)
 - `NewSDK` in `go.opentelemetry.io/contrib/config` now returns a configured SDK with a valid `MeterProvider`. (#4804)
+- The `go.opentelemetry.io/contrib/processors/baggage` module.
+  This module provides a Baggage Span Processor. (#5404)
```
```suggestion
- The `go.opentelemetry.io/contrib/processors/baggage` module. This module provides a Baggage Span Processor. (#5404)
```

pellared commented on processors/baggage/baggagetrace/go.mod:
```diff
@@ -0,0 +1,20 @@
+module go.opentelemetry.io/contrib/processors/baggage
```
```suggestion
module go.opentelemetry.io/contrib/processors/baggage/baggagetrace
```

pellared commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Add the new `go.opentelemetry.io/contrib/instrgen` package to provide auto-generated source code instrumentation. (#3068, #3108)
 - `NewSDK` in `go.opentelemetry.io/contrib/config` now returns a configured SDK with a valid `MeterProvider`. (#4804)
+- The `go.opentelemetry.io/contrib/processors/baggage` module. This module provides a Baggage Span Processor. (#5404)
```
```suggestion
- The `go.opentelemetry.io/contrib/processors/baggage/baggagetrace` module. This module provides a Baggage Span Processor. (#5404)
```

pellared commented on processors/baggage/baggagetrace/processor.go:
```diff
@@ -0,0 +1,42 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package baggagetrace // import "go.opentelemetry.io/contrib/processors/baggage"
```
```suggestion
package baggagetrace // import "go.opentelemetry.io/contrib/processors/baggage/baggagetrace"
```

pellared commented on processors/baggage/baggagetrace/doc.go:
```diff
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Package baggagetrace implements a baggage span processor.
+package baggagetrace // import "go.opentelemetry.io/contrib/processors/baggage"
```
```suggestion
package baggagetrace // import "go.opentelemetry.io/contrib/processors/baggage/baggagetrace"
```

codeboten commented on CODEOWNERS:
```diff
@@ -46,6 +46,8 @@ instrumentation/net/http/httptrace/otelhttptrace/                       @open-te
 instrumentation/net/http/otelhttp/                                      @open-telemetry/go-approvers @Aneurysm9 @dmathieu
 instrumentation/runtime/                                                @open-telemetry/go-approvers @MadVikingGod
 
+processors/baggage/                                                     @open-telemetry/go-approvers @codeboten
```
```suggestion
processors/baggage/baggagetrace                                         @open-telemetry/go-approvers @codeboten
```

pellared commented:
> I see this is approved, was there more work needed before merging this?

It was discussed during last SIG meetings. We just want to have more reviews.

MikeGoldsmith commented on CODEOWNERS:
```diff
@@ -46,6 +46,8 @@ instrumentation/net/http/httptrace/otelhttptrace/                       @open-te
 instrumentation/net/http/otelhttp/                                      @open-telemetry/go-approvers @Aneurysm9 @dmathieu
 instrumentation/runtime/                                                @open-telemetry/go-approvers @MadVikingGod
 
+processors/baggage/baggagetrace                                         @open-telemetry/go-approvers @codeboten
```
Happy to add my name to the component owners sections too.

```suggestion
processors/baggage/baggagetrace                                         @open-telemetry/go-approvers @codeboten @MikeGoldsmith
```

MikeGoldsmith commented:
> > I see this is approved, was there more work needed before merging this?
> 
> It was discussed during last SIG meetings. We just want to have more reviews.

Hey @pellared - what is the best way to get more community reviews? This PR has been open for more than two weeks and only a couple of approvers have provided feedback.

Do you have a sense how many reviews and/or how long you'd want to wait for additional reviews before accepting?

Is it specific people whom you would like to provide feedback?

MikeGoldsmith commented:
FYI the CI failure is from codecob being rate limited
```
debug - 2024-05-02 15:31:38,221 -- Commit creating result --- {"result": "RequestResult(error=RequestError(code='HTTP Error 429', params={}, description='{\"detail\":\"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 491 seconds.\"}'), warnings=[], status_code=429, text='{\"detail\":\"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 491 seconds.\"}')"}
error - 2024-05-02 15:31:38,221 -- Commit creating failed: {"detail":"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 491 seconds."}
```

pellared commented:
@MrAlias, if I remember correctly you wanted to review this PR as well.
